File: 0085-maximal-rectangle/0085-maximal-rectangle.cpp
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        int rows = matrix.size();
        int cols = matrix[0].size();
        int res = INT_MIN;

        vector<vector<int>> height(rows, vector<int>(cols, 0));


        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (i == 0) {
                    height[i][j] = (int) matrix[i][j] - '0';
                } else {
                    if (matrix[i][j] == '1') {
                        height[i][j] = height[i - 1][j] + 1;
                    } else {
                        height[i][j] = 0;
                    }
                }
                
            }
            
            int mx_area_in_row = largestRectangleArea(height[i]);
            res = max(res, mx_area_in_row);
        }

        return res;
    }


    int largestRectangleArea(vector<int>& heights) {
        int len = heights.size();

        vector<int> next_smaller_idx(len, len);
        vector<int> prev_smaller_idx(len, -1);

        stack<int> st;
        for (int i = len - 1; i >= 0; i--) {
            // loop till we have a lesser element on top or stack becomes empty.
            while (!st.empty())
            {
                // pop elements that aren't lesser than the current element
                if (heights[st.top()] >= heights[i]) {
                    st.pop();
                }
                // the next leser element is now on the top of the stack
                else {
                    next_smaller_idx[i] = st.top();
                    break;
                }
            }
    
            // push current element into the stack
            st.push(i);
        }

        while(!st.empty()) {
            st.pop();
        }


        for (int i = 0; i < len; i++) {
            // loop till we have a lesser element on top or stack becomes empty.
            while (!st.empty())
            {
                // pop elements that aren't lesser than the current element
                if (heights[st.top()] >= heights[i]) {
                    st.pop();
                }
                // the next leser element is now on the top of the stack
                else {
                    prev_smaller_idx[i] = st.top();
                    break;
                }
            }
    
            // push current element into the stack
            st.push(i);
        }

        int res = INT_MIN;
        for (int i = 0; i < len; i++) {
            int ar = 0;
            int left = prev_smaller_idx[i];
            int right = next_smaller_idx[i];

            ar = (heights[i] * (right - left - 1));

            res = max(res, ar);
        }

        return res;
    }




};

File: 0085-maximal-rectangle/0085-maximal-rectangle.java
import java.util.Stack;
import java.util.Arrays;

class Solution {
    public int maximalRectangle(char[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return 0;
        }

        int rows = matrix.length;
        int cols = matrix[0].length;
        int[][] height = new int[rows][cols];

        // Build height matrix
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (r == 0) {
                    height[r][c] = (matrix[r][c] == '1') ? 1 : 0;
                } else {
                    if (matrix[r][c] == '1') {
                        height[r][c] = 1 + height[r - 1][c];
                    } else {
                        height[r][c] = 0;
                    }
                }
            }
        }

        // Optional: print height matrix (debug)
        /*
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                System.out.print(height[r][c] + "\t");
            }
            System.out.println();
        }
        */

        int res = 0;
        for (int r = 0; r < rows; r++) {
            int[] prevSmaller = getPrevSmaller(height[r]);
            int[] nextSmaller = getNextSmaller(height[r]);

            for (int c = 0; c < cols; c++) {
                if (height[r][c] > 0) {
                    int prev = prevSmaller[c];
                    int next = nextSmaller[c];

                    int width = next - prev - 1;
                    int area = width * height[r][c];

                    res = Math.max(res, area);
                }
            }
        }

        return res;
    }

    private int[] getPrevSmaller(int[] a) {
        int n = a.length;
        int[] prevSmaller = new int[n];
        Arrays.fill(prevSmaller, -1);

        Stack<Integer> st = new Stack<>();
        for (int i = 0; i < n; i++) {
            while (!st.isEmpty() && a[st.peek()] >= a[i]) {
                st.pop();
            }

            if (!st.isEmpty()) {
                prevSmaller[i] = st.peek();
            }

            st.push(i);
        }

        return prevSmaller;
    }

    private int[] getNextSmaller(int[] a) {
        int n = a.length;
        int[] nextSmaller = new int[n];
        Arrays.fill(nextSmaller, n);

        Stack<Integer> st = new Stack<>();
        for (int i = 0; i < n; i++) {
            while (!st.isEmpty() && a[st.peek()] > a[i]) {
                int top = st.pop();
                nextSmaller[top] = i;
            }

            st.push(i);
        }

        return nextSmaller;
    }
}